File: src/main/java/net/dumbdogdiner/dogcore/commands/economy/EcoCommand.java
package net.dumbdogdiner.dogcore.commands.economy;

import dev.jorel.commandapi.annotations.Alias;
import dev.jorel.commandapi.annotations.Command;
import dev.jorel.commandapi.annotations.Permission;
import dev.jorel.commandapi.annotations.Subcommand;
import dev.jorel.commandapi.annotations.arguments.ALongArgument;
import dev.jorel.commandapi.annotations.arguments.AOfflinePlayerArgument;
import net.dumbdogdiner.dogcore.Permissions;
import net.dumbdogdiner.dogcore.database.User;
import net.dumbdogdiner.dogcore.messages.Messages;
import org.bukkit.OfflinePlayer;
import org.bukkit.command.CommandSender;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.CompletionStage;
import java.util.function.Function;

@Command("eco")
@Alias("economy")
@Permission(Permissions.ECO_ADMIN)
public final class EcoCommand {
    private EcoCommand() { }

    private static void economyHelper(
            final @NotNull CommandSender sender,
            final @NotNull OfflinePlayer player,
            final @NotNull Function<@NotNull User, @NotNull CompletionStage<@NotNull Boolean>> action
    ) {
        User.lookupCommand(player, sender, user -> action.apply(user).thenAccept(success -> {
            if (success) {
                sender.sendMessage(Messages.get("commands.economy.success"));
            } else {
                sender.sendMessage(Messages.get("error.failedTransaction"));
            }
        }));
    }

    @Subcommand("give")
    public static void give(
            final @NotNull CommandSender sender,
            final @NotNull @AOfflinePlayerArgument OfflinePlayer player,
            final @ALongArgument(min = 1) long amount
    ) {
        economyHelper(sender, player, user -> user.give(amount));
    }

    @Subcommand("take")
    public static void take(
            final @NotNull CommandSender sender,
            final @NotNull @AOfflinePlayerArgument OfflinePlayer player,
            final @ALongArgument(min = 1) long amount
    ) {
        economyHelper(sender, player, user -> user.give(-amount));
    }

    @Subcommand("set")
    public static void set(
            final @NotNull CommandSender sender,
            final @NotNull @AOfflinePlayerArgument OfflinePlayer player,
            final @ALongArgument(min = 0) long amount
    ) {
        economyHelper(sender, player, user -> user.setBalance(amount));
    }
}

File: src/main/kotlin/net/dumbdogdiner/dogcore/commands/EcoCommand.kt
package net.dumbdogdiner.dogcore.commands

import dev.jorel.commandapi.kotlindsl.anyExecutor
import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.longArgument
import dev.jorel.commandapi.kotlindsl.multiLiteralArgument
import dev.jorel.commandapi.kotlindsl.offlinePlayerArgument
import net.dumbdogdiner.dogcore.db.DbPlayer
import net.kyori.adventure.text.Component
import org.bukkit.OfflinePlayer

fun ecoCommand() = commandAPICommand("eco") {
    withAliases("economy")
    withPermission("dogcore.eco")

    multiLiteralArgument("give", "take", "set")
    offlinePlayerArgument("player")
    longArgument("amount", min = 0)

    anyExecutor { sender, args ->
        val player = DbPlayer.lookup(args["player"] as OfflinePlayer)
            ?: run {
                sender.sendMessage(Component.text("Player argument needed in this context."))
                return@anyExecutor
            }
        val amount = args["amount"] as Long
        val success = when (args[0]) {
            "give" -> player.give(amount)
            "take" -> player.give(-amount)
            "set" -> {
                player.balance = amount
                true
            }
            else -> return@anyExecutor
        }
        if (success) {
            sender.sendMessage(Component.text("The balance was updated"))
        } else {
            sender.sendMessage(Component.text("The balance would overflow, so it was not updated."))
        }
    }
}
